calico_containers/pycalico/block.py

```diff
@@ -0,0 +1,383 @@
+            ordinals.append(ordinal)
+            old_count = deleting_ref_counts.get(attr_idx, 0)
+            deleting_ref_counts[attr_idx] = old_count + 1
+
+        # Compute which attributes need to be cleaned up.  We do this by
+        # reference counting.  If we're deleting all the references, then it
+        # needs to be cleaned up.
+        attr_index_to_delete = set()
+        ref_counts = self._get_attribute_ref_counts()
+        for idx, refs in deleting_ref_counts.iteritems():
+            if ref_counts[idx] == refs:
+                attr_index_to_delete.add(idx)
+
+        # Delete attributes if necessary
+        if attr_index_to_delete:
+            self._delete_attributes(attr_index_to_delete, ordinals)
+
+        # All attributes updated.  Finally, release all the requested
+        # addressses.
+        for ordinal in ordinals:
+            self.allocations[ordinal] = None
+
+        return unallocated
+
+    def _delete_attributes(self, attr_index_to_delete, ordinals):
+        """
+        Delete some attributes (used during release processing).
+
+        :param attr_index_to_delete: set of indexes of attributes to delete
+        :param ordinals: list of ordinals of IPs to release (for debugging)
+        :return: None.
+        """
+        new_indexes = range(len(self.attributes))
+        new_attributes = []
+        y = 0  # next free slot in new attributes list.
+        for x in xrange(len(self.attributes)):
+            if x in attr_index_to_delete:
+                # current attr at x being deleted.
+                new_indexes[x] = None
+            else:
+                # current attr at x is kept.
+                new_indexes[x] = y
+                y += 1
+                new_attributes.append(self.attributes[x])
+        self.attributes = new_attributes
+
+        # Spin through all the allocations and update indexes
+        for i in xrange(BLOCK_SIZE):
+            if self.allocations[i] is not None:
+                new_index = new_indexes[self.allocations[i]]
+                self.allocations[i] = new_index
+                # If the new index is None, we better be releasing that
+                # address
+                assert new_index is not None or i in ordinals
+
+    def _get_attribute_ref_counts(self):
+        """
+        Walk the allocations and get a dictionary of reference counts to each
+        set of attributes.
+        """
+        ref_counts = {}
+        for a in self.allocations:
+            old_counts = ref_counts.get(a, 0)
+            ref_counts[a] = old_counts + 1
+        return ref_counts
+
+    def _find_or_add_attrs(self, primary_key, attributes):
+        """
+        Check if the key and attributes match existing and return the index, or
+        if they don't exist, add them and return the index.
+        """
+        assert json.dumps(attributes), \
+            "Attributes aren't JSON serializable."
+        attr = {AllocationBlock.ATTR_PRIMARY: primary_key,
```

Comment:
After our chat I find this naming somewhat confusing.  Primary key just sounds like it should be a unique id of the attribute set to me.